src/Cardano/CoinSelection/Random.hs

```diff
@@ -49,66 +49,168 @@ import Data.Functor
--- of the same value `x`, then we will end up with a lot of change outputs in
--- our UTxO of size `x` precisely when we have a lot of payment requests of
--- size `x`
--- @
---
--- @
--- **Self organisation principle 3**
--- Searching the UTxO for additional entries to improve our change output is
--- only useful if the UTxO contains entries that are sufficiently small enough.
--- But precisely when the UTxO contains many small entries, it is less likely
--- that a randomly chosen UTxO entry will push the total above the upper bound
--- we set.
--- @
+-- | An implementation of the __Random-Improve__ coin selection algorithm.
+--
+-- = Overview
+--
+-- The __Random-Improve__ coin selection algorithm works in __two phases__, by
+-- /first/ selecting UTxO entries /at random/ to pay for each of the given
+-- outputs, and /then/ attempting to /improve/ upon each of the selections.
+--
+-- === Phase 1: Random Selection
+--
+-- __In this phase, the algorithm randomly selects a minimal set of UTxO__
+-- __entries to pay for each of the given outputs.__
+--
+-- During this phase, the algorithm:
+--
+--   *  processes outputs in /descending order of coin value/.
+--
+--   *  maintains a /remaining UTxO set/, initially equal to the given
+--      /UTxO set/ parameter.
+--
+-- For each output of value __/v/__, the algorithm /randomly/ selects entries
+-- from the /remaining UTxO set/, until the total value of selected entries is
+-- greater than or equal to __/v/__. The selected entries are then associated
+-- with that output, and removed from the /remaining UTxO set/.
+--
+-- This phase ends when every output has been associated with a selection of
+-- UTxO entries.
+--
+-- However, if the remaining UTxO set is completely exhausted before all
+-- outputs can be processed, the algorithm terminates and delegates to the
+-- __Largest-First__ algorithm. (See 'largestFirst'.)
+--
+-- === Phase 2: Improvement
+--
+-- __In this phase, the algorithm attempts to improve upon each of the UTxO__
+-- __selections made in the previous phase, by conservatively expanding the__
+-- __selection made for each output.__
+--
+-- During this phase, the algorithm:
+--
+--   *  processes outputs in /ascending order of coin value/.
+--
+--   *  continues to maintain the /remaining UTxO set/ produced by the previous
+--      phase.
+--
+--   *  maintains an /accumulated coin selection/, which is initially /empty/.
+--
+-- For each output of value __/v/__, the algorithm:
+--
+--  1.  __Calculates a /target range/__ for the total value of inputs used to
+--      pay for that output, defined by the triplet:
+--
+--      (/minimum/, /ideal/, /maximum/) = (/v/, /2v/, /3v/)
+--
+--  2.  __Attempts to /improve/ upon the /existing UTxO selection/__ for that
+--      output, by repeatedly selecting additional entries at random from the
+--      /remaining UTxO set/, stopping when the selection can be improved upon
```

Comment:
you mean `cannot be further improved`, right?

Reply:
Yes, that's precisely the meaning!

This wording is actually okay.

We could say "cannot be further improved upon", but in this context, the original wording seems just as natural.

The "can be _verb_-ed no further" phrase is similar to "can _verb_ no further".

Similar examples:

> "We stopped walking when our legs could carry us no further."
> "After we reached an obstacle in the road, we could drive no further."
> "The algorithm stops when the counter can be decreased no further."
> "The sorting procedure halts when the the list can be shortened no further."
> "We continue until the metal reaches absolute zero and can be cooled no further."